core/src/main/java/com/linecorp/armeria/client/DefaultDnsQueryLifecycleObserver.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client;
+
+import java.net.InetSocketAddress;
+import java.util.List;
+
+import io.micrometer.core.instrument.MeterRegistry;
+import io.netty.channel.ChannelFuture;
+import io.netty.handler.codec.dns.DnsQuestion;
+import io.netty.handler.codec.dns.DnsResponseCode;
+import io.netty.resolver.dns.DnsQueryLifecycleObserver;
+
+/**
+ * A {@link DnsQueryLifecycleObserver} that helps capture custom dns metrics.
+ */
+public class DefaultDnsQueryLifecycleObserver implements DnsQueryLifecycleObserver {
+
+    private final MeterRegistry meterRegistry;
+    private final DnsQuestion dnsQuestion;
+
+    /**
+     * Accepts meterRegistry.
+     * @param meterRegistry {@link MeterRegistry} MeterRegistry to capture metrics.
+     * @param question {@link DnsQuestion} DnsQuestion.
+     */
+    public DefaultDnsQueryLifecycleObserver(MeterRegistry meterRegistry, DnsQuestion question) {
+        this.dnsQuestion = question;
+        this.meterRegistry = meterRegistry;
+    }
+
+    @Override
+    public void queryWritten(InetSocketAddress dnsServerAddress, ChannelFuture future) {
+        meterRegistry.counter("dns.query.count", dnsQuestion.name(), dnsQuestion.type().name()).increment();
```

Comment:
- How about creating `Counter` in the constructor and reuse it?
- I think we can use `failed` and `succeeded` as a tag. Then we don't need to count for total queries. 

You can refer the follwing code:
https://github.com/line/armeria/blob/9061e241961c806951dd855be65fd871131a1835/core/src/main/java/com/linecorp/armeria/internal/common/metric/RequestMetricSupport.java#L258-L260

Reply:
On second thought `queryWritten` has meaning it self. How about add a tag with `written`?

Reply:
Sounds good I will make the corresponding changes.